### PR TITLE
Librispeech Deepspeech JAX workload 

### PR DESCRIPTION
**Pull request by sourabh2k15**
this is the final version that doesn't hang 
this version has grad_clip logic commented as that causes workload to hang on GPUs , we'll fix it in a future PR for e2e run

### CODE REVIEW

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+    Returns:
+      The encoded sequence of inputs, shaped <float32>[batch_size, seq_len,
+        hidden_size], as well as the final hidden states of the RNN cell. For an
+        LSTM cell the final states are a tuple (c, h), each shaped <float32>[
+          batch_size, hidden_size].
+    """
+    if reverse:
+      inputs = flip_sequences(inputs, lengths)
+
+    recurrent_dropout_mask = None
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:436]: if this is always `None`, can we fully delete it?

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+            inputs,
+            lengths,
+            initial_states=initial_states,
+            deterministic=deterministic)
+
+
+class BatchRNN(nn.Module):
+  """Implements a single conformer encoder layer.
+
+  High level overview:
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:640]: is this missing some docstring, or should it be deleted?

File: submission_runner.py [hunk 3]
```diff
@@ -263,13 +262,6 @@ def train_once(
             rng=update_rng)
     except spec.TrainingCompleteError:
       training_complete = True
-    except RuntimeError as e:
-      if "out of memory" in str(e):
-        logging.warning(
-            f'error: GPU out of memory during training during step {global_step}, error: {str(e)}'  # pylint: disable=line-too-long
-        )
-        if torch.cuda.is_available():
-          torch.cuda.empty_cache()
```
Comment (znado) [inline submission_runner.py:272]: should we keep this?

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/workload.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import flax.linen as nn
+import jax
+from jax import lax
+import jax.numpy as jnp
+import numpy as np
+import optax
+
+from algorithmic_efficiency import spec
+from algorithmic_efficiency.workloads.librispeech_conformer.librispeech_jax.workload import \
+    LibriSpeechConformerWorkload
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/workload.py:18]: do we need to check if model_state is not none here? https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py#L61

also do we need to return model_state here instead of None? https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py#L82

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  del model_state
+  del rng
+
+  params_zeros_like = jax.tree_map(lambda s: jnp.zeros(s.shape_tuple),
+                                   workload.param_shapes)
+  opt_init_fn, opt_update_fn = optimizer(hyperparameters,
+                                         workload.num_train_examples)
+  optimizer_state = opt_init_fn(params_zeros_like)
+  # log_pytree_shape_and_statistics(
+  #     flax.jax_utils.unreplicate(model_params), workload.summary_writer)
```
Comment (znado) [inline reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:51]: delete

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    logging.info('Empty pytree')
+    return
+  logging.info('Printing model param shapes.')
+  shape_dict = jax.tree_map(_summary_str, pytree)
+
+  logging.info(shape_dict.pretty_repr())
+
+  total_params = jax.tree_util.tree_reduce(
+      operator.add, jax.tree_map(lambda x: x.size, pytree))
+  print('num params = ', total_params)
```
Comment (znado) [inline reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:98]: delete

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  (loss, new_model_state), grad = grad_fn(params)
+
+  grad_clip = hyperparameters.grad_clip
+  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
+
+  grad_scaling_factor = grad_clip / (grad_norm + _GRAD_CLIP_EPS)
+  grad_scaling_factor = jax.lax.clamp(min=0.0, x=grad_scaling_factor, max=1.0)
+
+  grad = jax.tree_map(lambda x: x * grad_scaling_factor, grad)
+  loss, grad = lax.pmean((loss, grad), axis_name='batch')
```
Comment (znado) [inline reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:137]: I don't remember what we settled on, but is it possible to pmean before calculating grad_norm? if not, can you add a comment saying we need pmean after otherwise the code hangs?

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+      update_fn, axis_name='batch', in_axes=(0, 0, 0, 0))
+  new_params, new_model_state, new_optimizer_state, loss, grad_norm = pmapped_update_step(  # pylint: disable=line-too-long
+    batch,
+    current_param_container,
+    model_state,
+    optimizer_state)
+
+  print('{}) loss = {}, grad_norm = {}'.format(global_step,
+                                               loss.mean(),
+                                               grad_norm.mean()))
```
Comment (znado) [inline reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:183]: delete

**Review by znado — CHANGES_REQUESTED** [review]

File: algorithmic_efficiency/workloads/librispeech_conformer/workload.py [hunk 9]
```diff
@@ -37,7 +37,7 @@ def num_train_examples(self):
 
   @property
   def num_eval_train_examples(self):
-    return 512
+    return 512
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_conformer/workload.py]: done

File: algorithmic_efficiency/workloads/librispeech_conformer/workload.py [hunk 10]
```diff
@@ -157,7 +167,7 @@ def _build_dataset(self,
     train = False
 
     if split == 'train':
-      split = 'train-clean-100+train-clean-360+train-other-500'
+      split = 'train-clean-100'
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_conformer/workload.py:166]: done

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 11]
```diff
@@ -0,0 +1,732 @@
...
+  batch_norm_momentum: float = 0.999
+  batch_norm_epsilon: float = 0.001
+  input_dropout_rate: float = 0.1
+  feed_forward_dropout_rate: float = 0.1
+  enable_residual_connections: bool = True
+  enable_decoder_layer_norm: bool = True
+  bidirectional: bool = True
+
+
+class Subsample(nn.Module):
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:60]: okay for now I'll leave it unmerged so whoever's debugging slowness in deepspeech can only look at this file for model code

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 12]
```diff
@@ -0,0 +1,732 @@
...
+
+    # Mask outputs by correct paddings to ensure padded elements in inputs map
+    # to padded value in outputs.
+    outputs = outputs * (1.0 -
+                         jnp.expand_dims(jnp.expand_dims(out_padding, -1), -1))
+
+    return outputs, out_padding
+
+
+class FeedForwardModule(nn.Module):
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:182]: not touching deepspeech model code for now , will update later

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 13]
```diff
@@ -0,0 +1,732 @@
...
+            inputs)
+    inputs *= padding_mask
+
+    inputs = nn.Dropout(rate=config.feed_forward_dropout_rate)(
+        inputs, deterministic=not train)
+
+    return inputs
+
+
+class LayerNorm(nn.Module):
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:208]: will merge common parts between conformer and deepspeech in a future PR

File: submission_runner.py (hunk 3 above)
Comment (sourabh2k15) [reply submission_runner.py:272]: it's extra code that doesn't really catch OOM errors, I included it last time in hope that it might catch a real GPU OOM error

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py (hunk 8 above)
Comment (sourabh2k15) [reply reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:183]: converted to logging.info() 
we need this to have visibility into the run

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py (hunk 7 above)
Comment (sourabh2k15) [reply reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:137]: so with batch size 256 doesn't matter if grad clip code is before / after pmean it still hangs 
I've commented out the grad_clip code and I'm unsure if this workload even needs gradient clipping , we never ran it without clip 

for now I've kept commented code and we can figure out if we need it / how to make it not hang if we do need it

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/workload.py (hunk 4 above)
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/workload.py:18]: 1) it will fail automatically if batch_stats is not passed in 
2) we return model_state as None for eval mode and we ignore that returned model_state in eval code path so it's fine

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py (hunk 6 above)
Comment (sourabh2k15) [reply reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:98]: done

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/e2e_test.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
+from algorithmic_efficiency.workloads.librispeech_deepspeech.librispeech_jax import \
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/e2e_test.py]: moved to tests/ 
I still want to keep it checked in for whoever debugs this workload slowness

File: algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py [hunk 15]
```diff
@@ -6,12 +6,11 @@
...
 import tensorflow as tf
 
-from algorithmic_efficiency import spec
-
+tf.config.set_visible_devices([], 'GPU')
 
 def get_librispeech_dataset(split_name: str,
                             data_dir: str,
-                            shuffle_rng: spec.RandomState,
+                            shuffle_rng,
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py:14]: done

File: algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py [hunk 16]
```diff
@@ -6,12 +6,11 @@
 import numpy as np
 import tensorflow as tf
 
-from algorithmic_efficiency import spec
-
+tf.config.set_visible_devices([], 'GPU')
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py:9]: removed

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+      update_step,
+      opt_update_fn=opt_update_fn,
+      global_step=global_step,
+      workload=workload,
+      hyperparameters=hyperparameters,
+      rng=rng,
+      lr=lr)
+
+  pmapped_update_step = jax.pmap(
+      update_fn, axis_name='batch', in_axes=(0, 0, 0, 0))
```
Comment (znado) [inline reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:153]: we need to refactor this to avoid pmap inside the loop like in conformer, right? https://github.com/mlcommons/algorithmic-efficiency/blob/main/reference_submissions/librispeech_conformer/librispeech_jax/submission.py#L80

File: setup.cfg [hunk 18]
```diff
@@ -36,7 +36,7 @@ setup_requires =
 # Dependencies of the project:
 install_requires =
   absl-py==1.0.0
-  numpy==1.21.6
+  numpy==1.23.3
```
Comment (znado) [inline setup.cfg:39]: we need to undo this change, I think it'll break on lower versions of python

File: tests/e2e_test.py [hunk 19]
```diff
@@ -0,0 +1,230 @@
+# pylint: skip-file
```
Comment (znado) [inline tests/e2e_test.py:1]: I think we should hold off on checking this in for now (but definitely still keep it around in your fork!!).

at some point we'll need to write up serious e2e testing for _all_ workloads, so we'll have this functionality then anyways

Comment (znado) [reply tests/e2e_test.py:1]: (also removing this should fix the remaining lint issue)

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 20]
```diff
@@ -0,0 +1,730 @@
...
+
+  Args:
+    inputs: An array of input IDs <int>[batch_size, seq_length].
+    lengths: The length of each sequence <int>[batch_size].
+
+  Returns:
+    An ndarray with the flipped inputs.
+  """
+  # Compute the indices to put the inputs in flipped order as per above example.
+  max_length = inputs.shape[0]
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:349]: instead of taking `lengths` as an input, I think we can recompute it as `lengths = jnp.argmin(inputs, axis=1)`

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py (hunk 1 above)
Comment (znado) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:436]: I still think we should delete this, but we can leave it for a future cleanup

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 21]
```diff
@@ -0,0 +1,730 @@
...
+    if reverse:
+      inputs = flip_sequences(inputs, lengths)
+
+    recurrent_dropout_mask = None
+    _, (cell_states, outputs) = self.unroll_cell(initial_state,
+                                                 inputs,
+                                                 recurrent_dropout_mask,
+                                                 deterministic)
+    final_state = jax.tree_map(
+        lambda x: x[jnp.arange(inputs.shape[0]), lengths - 1], cell_states)
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:442]: as far as I can tell, we never use this, so lets delete this code

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 22]
```diff
@@ -0,0 +1,730 @@
...
+  def __call__(self, inputs, input_paddings, train):
+    config = self.config
+
+    inputs = BatchNorm(config.encoder_dim,
+                       config.dtype,
+                       config.batch_norm_momentum,
+                       config.batch_norm_epsilon)(inputs, input_paddings, train)
+    lengths = jnp.sum(1 - input_paddings, axis=-1, dtype=jnp.int32)
+
+    output, _ = LSTM(
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:652]: return just outputs and delete the `, _`

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 23]
```diff
@@ -0,0 +1,730 @@
...
+      The sequence of all outputs for the final layer, and a list of final
+      states for each cell and direction. Directions are alternated (first
+      forward, then backward, if bidirectional). For example for a bidirectional
+      cell this would be: layer 1 forward, layer 1 backward, layer 2 forward,
+      layer 2 backward, etc..
+      For some cells like LSTMCell a state consists of an (c, h) tuple, while
+      for others cells it only contains a single vector (h,).
+    """
+    batch_size = inputs.shape[0]
+    final_states = []
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:509]: we never use `final_states`, so we should be able to delete this

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+import jax
+from jax import lax
+import jax.numpy as jnp
+import optax
+
+from algorithmic_efficiency import spec
+
+# import gc
+# import torch
+
```
Comment (znado) [inline reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:16]: delete?

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+        spec.ForwardPassMode.TRAIN,
+        {'params' : params_rng, 'dropout' : dropout_rng})
+
+    loss = workload.loss_fn(batch['targets'], (logits, logit_paddings))
+    return loss, new_batch_stats
+
+  grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+  (loss, new_model_state), grad = grad_fn(params)
+
+  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
```
Comment (znado) [inline reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:107]: (we can follow up offline too, but)

is it possible to flip the order of these lines?

```
  loss, grad = lax.pmean((loss, grad), axis_name='batch')
  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
```

**Review by znado — CHANGES_REQUESTED** [review]

File: tests/e2e_test.py (hunk 19 above)
Comment (sourabh2k15) [reply tests/e2e_test.py]: sounds good, deleted

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py (hunk 25 above)
Comment (sourabh2k15) [reply reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py]: fixed it in latest deepspeech submission code, now pmean happens before grad clipping

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py (hunk 24 above)
Comment (sourabh2k15) [reply reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py]: deleted

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py (hunk 23 above)
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py]: deleted final_states

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py (hunk 22 above)
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py]: deleted

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py (hunk 21 above)
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py]: deleted

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py (hunk 20 above)
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:350]: made the change to pass input_paddings as deep as possible

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 26]
```diff
@@ -0,0 +1,730 @@
...
+
+    return outputs, output_paddings
+
+
+@jax.jit
+def hard_tanh(x, min_value, max_value):
+  return jnp.where(x < min_value,
+                   min_value,
+                   jnp.where(x > max_value, max_value, x))
+
```
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py]: removed

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py (hunk 17 above)
Comment (sourabh2k15) [reply reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py]: that's correct, updated pmap code to be same as conformer submission.py

File: reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py (hunk 7 above)
Comment (sourabh2k15) [reply reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py]: resolved in newest version of code, pmean happens before grad clipping now 
I've still eliminated jax.lax.cond in our grad clipping code as that's more efficient

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/e2e_test.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/e2e_test.py]: removed it for now

**Comment by sourabh2k15** [comment]
update : this PR is good to go, here's the e2e run : https://tensorboard.dev/experiment/45h1nSXfRqypQdgpN6Dgyw 

step time looks good, will check tomorrow if it achieves expected target validation WER 

changes done to model in final state :
1) removed a BatchNorm in convolution subsampling layer which is not present in conformer and is unclear if present in original paper , this specific BatchNorm causes a 2 minute constant folding compilation stage on GPU causing slowdown 
2) fixed a bug in  FeedForwardModule , we were missing relu activation function 
3) add cross device syncing code in BatchNorm so mean and variance are computed from the entire batch not just sharded batch 
4) put grad_clip = 5.0 given initial gradient norms do go higher from tensorboard 

changes done to submission code to resolve recompilation : 
pmapped train step was written in a way that triggers recompilation at every step, this was resolved by adopting the pmap code from conformer submission file 

debugging note : can use JAX_LOG_COMPILES=1 to see whenever jax recompiles stuff

**Review by znado — APPROVED** [review]
Looks great, it's fantastic to finally get this in!
